File: examples/node/example.test.ts
/**
 * leak-assert — Node.js / TypeScript usage examples
 * Run with: node --expose-gc node_modules/.bin/jest
 */
import { LeakTest, kb, mb } from 'leak-assert'

// ── Example 1: Simple map that leaks ─────────────────────────────────────────

const leakyCache = new Map<string, Buffer>()

test('leaky cache is caught', async () => {
  const lt = new LeakTest({ warmup: 50, iterations: 500 })

  await lt.runAsync(async () => {
    // Bug: never evicts — this will fail the growth_rate assertion
    leakyCache.set(Math.random().toString(), Buffer.alloc(1024))
  })

  // Expect this test to FAIL — demonstrates the framework catching a leak
  expect(() => {
    lt.assert({ growthRate: kb(0.5) })
  }).toThrow('leak-assert [growthRate]')
})

// ── Example 2: Well-behaved function passes ───────────────────────────────────

test('clean function passes memory assertions', async () => {
  const lt = new LeakTest({
    name:       'clean-handler',
    warmup:     100,
    iterations: 1000,
    gc:         { force: true, between: 'sample' },
  })

  await lt.runAsync(async () => {
    // allocate and release — no retention
    const buf = Buffer.alloc(4096)
    buf.fill(0)
    // buf goes out of scope, GC can collect
  })

  lt.printSummary()

  lt.assert({
    growthRate: kb(1),   // < 1kb per iteration
    stable:     mb(5),   // heap returns within 5MB of start
    ceiling:    mb(200), // never exceeds 200MB absolute
  })
})

// ── Example 3: Async HTTP handler simulation ──────────────────────────────────

test('async handler does not leak event listeners', async () => {
  const { EventEmitter } = await import('events')
  const emitter          = new EventEmitter()

  const lt = new LeakTest({ iterations: 500, warmup: 50 })

  await lt.runAsync(async () => {
    const handler = () => {}
    emitter.on('data', handler)
    emitter.emit('data', { payload: 'test' })
    emitter.off('data', handler) // correct cleanup
  })

  lt.assert({
    growthRate: kb(1),
    stable:     mb(5),
  })
})

File: examples/go/example_test.go
package example_test

import (
	"testing"

	leakassert "github.com/leak-assert/leak-assert-go"
)

// ── Example 1: Leaky map is caught ───────────────────────────────────────────

var leakyMap = make(map[int][]byte)

func TestLeakyMapIsCaught(t *testing.T) {
	i := 0
	lt := leakassert.New(t, leakassert.Config{
		Warmup:     50,
		Iterations: 500,
	})

	lt.Run(func() {
		// Bug: never deletes — map grows forever
		leakyMap[i] = make([]byte, 1024)
		i++
	})

	// This assertion WILL fail — demonstrating leak detection
	// In real tests, remove the t.Skip and expect the assertion to fail CI
	t.Skip("intentional leak demo — remove skip to see failure")
	lt.Assert(leakassert.GrowthRate("512"))
}

// ── Example 2: Clean handler passes ─────────────────────────────────────────

func TestCleanHandlerNoLeak(t *testing.T) {
	lt := leakassert.New(t, leakassert.Config{
		Warmup:     100,
		Iterations: 2000,
		ForceGC:    true,
	})

	lt.Run(func() {
		// Allocate and release — GC will collect
		buf := make([]byte, 4096)
		_ = len(buf)
	})

	lt.PrintSummary()

	lt.Assert(
		leakassert.GrowthRate("1kb/iter"),
		leakassert.Stable(5*leakassert.MB),
		leakassert.Ceiling(200*leakassert.MB),
	)
}

// ── Example 3: Goroutine leak detected ──────────────────────────────────────

func TestNoGoroutineLeak(t *testing.T) {
	lt := leakassert.New(t, leakassert.Config{
		Iterations: 200,
		Warmup:     20,
	})

	lt.Run(func() {
		// Correctly waits for goroutine — no leak
		done := make(chan struct{})
		go func() {
			close(done)
		}()
		<-done
	})

	lt.Assert(
		leakassert.GrowthRate("1kb/iter"),
		leakassert.GoroutinesStable(),
	)
}
